reset_tstamp.8
.\" section 8 man page for reset_tstamp.sh
.\"
.\" This man page was first written by Cody Boone Ferguson for the IOCCC
.\" in 2022.
.\"
.\" Humour impairment is not virtue nor is it a vice, it's just plain
.\" wrong: almost as wrong as JSON spec misfeatures and C++ obfuscation! :-)
.\"
.\" "Share and Enjoy!"
.\"     --  Sirius Cybernetics Corporation Complaints Division, JSON spec department. :-)
.\"
.TH reset_tstamp.sh 8 "22 October 2022" "reset_tstamp.sh" "IOCCC tools"
.SH NAME
reset_tstamp \- reset the minimum timestamp \fBMIN_TIMESTAMP\fP
.SH SYNOPSIS
\fBreset_tstamp.sh\fP [\-h] [\-V] [\-v level] [\-l limit_ioccc.h]
.SH DESCRIPTION
\fBreset_tstamp.sh\fP will reset the minimum timestamp allowed for entries to the contest.
.SH OPTIONS
.TP
\fB\-h\fP
Show help and exit.
.TP
\fB\-V\fP
Print version and exit.
.TP
\fB\-v\fP
Set verbosity level.
.TP
\fB\-l\fP
Specify limit file.
.SH EXIT STATUS
.PP
An exit code of 0 indicates that the timestamp was successfully updated.
.PP
An exit code of 1 means Phase 0 of verification failed.
.PP
An exit code of 2 means Phase 1 of verification failed.
.PP
An exit code of 3 means Phase 2 of verification failed.
.PP
An exit code of 4 means Phase 3 of verification failed.
.PP
An exit code of 5 means Phase 4 of verification failed.
.PP
An exit code of 6 means \fBrpl(1)\fP failed to modify the \fIlimit_ioccc.h\fP file.
.PP
An exit code of 7 means that the limit file not found or not readable and not writable.
.PP
An exit code of 8 means \fB\-h\fP was used or \fB\-V\fP was used.
.PP
An exit code of 9 means a command line usage error.
.PP
An exit code >= 10 means an internal error occurred.
.SH NOTES
.PP
Only run this script when you wish to invalidate all timestamps prior to now, such as when you make a fundamental change to a critical JSON format, or make a fundamental change to the compressed tarball file structure, or make a critical change to limit_ioccc.h that is MORE restrictive.
.PP
\fBDO NOT\fP run this script simply for a new IOCCC!
.PP
Yes, we make it very hard to run this script for good reason.
Only IOCCC judges can perform ALL the steps needed to complete this action.


CHANGES.md
# Major changes to the IOCCC entry toolkit


## Release 0.6 2022-09-02

Released IOCCC entry toolkit v0.6 2022-09-02

Updated CHANGES.md for v0.6 2022-09-02.

Changed `MIN_TIMESTAMP` from 1655958810 to 1662145368.

Updated `mkiocccentry` from "0.40 2022-03-15" to "0.41 2022-09-02".
Updated `.info.json` version from "1.10 2022-06-22" to "1.11 2022-09-02".
Updated `.author.json` version from "1.13 2022-06-22" to "1.14 2022-09-02".

Improved code to use new facilities for output to a buffer from dbg release of
v2.5 2022-07-23.

The `chk_foo()` functions in `chk_validate.c` and the `test_foo()` functions in
`entry_util.c` are code complete, although they remain untested and unused.  The
`chkentry` tool, however, is not yet code complete. Later releases of tested
JSON semantic code will no doubt modify these functions.

Improved a number of the ways that JSON field values are checked.  In a number
of cases, code form `mkiocccentry.c` was moved into `test_foo()` functions so
that they could be used by other tools such as the JSON semantic test code.

Added code to generate JSON semantic tables from JSON reference files for
`.info.json` and `.author.json`.  The `jsemcgen.sh` tool manages this by way of
the `jsemtblgen` code generator and header, patch and trailer files (see
`chk.auth.*` and `chk.info.*` files).

Avoided the appearance of attacking any particular individual.  It was not our
intention to disrespect anyone, even though we disagree with some of the
technical decisions.  Where we have fundamental technical disagreements, we
attempted to express those technical disagreements with humour in hopefully a
more fun way.  As also now apologize for how `bison` and `flex` generated code
may look, instead of simply calling it ugly.  As such, we hopefully improved
some of the humour in our code while trying to be nice and friendly to others.

For example now the adjusted dbg levels in JSON parser are:

```
*     At -J 3, only the top level return type and top level tree are printed.
*     At -J 5, intermediate tree return types and tree are printed too.
*     At -J 7, also print grammar progress.
*     At -J 9, also print sorry_text and sorry_lang grammar values.
```

Removed a number of files and added a number of files under the `test_JSON/`
tree.  When the JSON semantic code is being tested in a future release, we
expect more such `test_JSON/` tree changes.

Improved / add a number of man pages.  Updated `README.md`.

Improved and expanded `txzchk`.

Added more test code.  We attempt to detect feathers in tarballs.  :-)

We will neither confirm nor deny the presence of an "Easter egg".
To to so would be "foolish".  :-)

Improved and fixed `vermod.sh` and `reset_tstamp.sh`.  Tested this
code by changing the `MIN_TIMESTAMP` as noted above.  The `MIN_TIMESTAMP`
needed to up updated anyway due to changes in the `.info.json` and
`.author.json` formats.

Made numerous improvements and bug fixes to the Makefile.

Fixed how `picky` is used in by the `make picky` rule for a few special files.

Added multiple rules to the Makefile including but not limited to `make mkref`,
`make legacy_clobber`, and `make legacy_clean` rules.  Applied multiple bug
fixes to the Makefile.

Improved the Makefile to be less impacting for modern systems while trying to
maintain, for as long as we can, compatibility with some older systems.

Attempted to improve compatibility with reasonably modern systems.  We managed
to keep CentOS 7 somewhat supported for now, although we may be forced to drop
support of such an old system before 2024. But if we do reach 2024 we can only
say that given that CentOS 7 will reach EOL in 2024 and CentOS 8 support was
unduly dropped and since < CentOS 7 already was at EOL this should not be a
problem. :-)

Users of very out of date systems can still enter and submit entries to the
IOCCC.  They just might need to find a more modern system to package and submit
their IOCCC entry, however.

Added stub code for `hostchk.sh`.  A future release will include mode tests for
the given hosts.  Future releases will also include a bug report system that
will also use `hostchk.sh`.

Improved the `no-comment` directive in `.info.json` and `.author.json` files.

Improved how `time_t` values are used and printed.  We no longer assume that
`time_t` is signed nor assume it is unsigned: we only assume it's time. :-)

Improved comments in C code about special `seqcexit` comment directives.

Make numerous bug fixes and fixed a fair number of typos.  Nevertheless we claim
this is *NOT* complete. :-)


## Release 0.5 2022-07-10

The `jsemtblgen` tool is code complete.

Added `is_reserved()` function to `rule_count.c`.

Changed the Official iocccsize version from 28.10 2022-03-15 to 28.11
2022-07-10. Updated `test_JSON` tree and `json_teststr.txt` to reflect the new
`IOCCCSIZE_VERSION` value.

Added and improved a number of man pages.

Added `Makefile` rule to build HTML files from all man pages.

Both the `parse_json()` and `parse_json_file()` top level functions of the JSON
parser return a JSON node parse tree when the JSON is valid.

Trimmed the JSON parser to not include code that does not strictly belong to the
generic JSON parser.  Removed dead C code.

Updated dbg facility to version 2.4 2022-06-26.  Version 2.4 2022-06-26 contains
functions to write (form) a message, debug diagnostic message, warning, or error
message into a string buffer:

```c
extern void snmsg(char *str, size_t size, char const *fmt, ...)
extern void sndbg(char *str, size_t size, int level, char const *fmt, ...);
extern void snwarn(char *str, size_t size, char const *name, char const *fmt, ...);
extern void snwarnp(char *str, size_t size, char const *name, char const *fmt, ...);
extern void snwerr(int error_code, char *str, size_t size, char const *name, char const *fmt, ...);
extern void snwerrp(int error_code, char *str, size_t size, char const *name, char const *fmt, ...);
```


reset_tstamp.sh
#!/usr/bin/env bash
#
# reset_tstamp.sh - reset the minimum timestamp: MIN_TIMESTAMP
#
# Only run this script when you wish to invalidate all timestamps prior to now,
# such as when you make a fundamental change to a critical JSON format, or make
# a fundamental change to the compressed tarball file structure, or make a
# critical change to limit_ioccc.h that is MORE restrictive.
#
# DO NOT run this script simply for a new IOCCC!
#
# Yes, we make it very hard to run this script for good reason.  Only IOCCC
# judges can perform ALL the steps needed to complete this action.
#
# Copyright (c) 2022 by Landon Curt Noll.  All Rights Reserved.
#
# Permission to use, copy, modify, and distribute this software and its
# documentation for any purpose and without fee is hereby granted, provided that
# the above copyright, this permission notice and text this comment, and the
# disclaimer below appear in all of the following:
#
#       supporting documentation
#       source copies
#       source works derived from this source
#       binaries derived from this source or from derived source
#
# LANDON CURT NOLL DISCLAIMS ALL WARRANTIES WITH REGARD TO THIS SOFTWARE,
# INCLUDING ALL IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS. IN NO
# EVENT SHALL LANDON CURT NOLL BE LIABLE FOR ANY SPECIAL, INDIRECT OR
# CONSEQUENTIAL DAMAGES OR ANY DAMAGES WHATSOEVER RESULTING FROM LOSS OF
# USE, DATA OR PROFITS, WHETHER IN AN ACTION OF CONTRACT, NEGLIGENCE OR
# OTHER TORTIOUS ACTION, ARISING OUT OF OR IN CONNECTION WITH THE USE OR
# PERFORMANCE OF THIS SOFTWARE.
#
# chongo (Landon Curt Noll, http://www.isthe.com/chongo/index.html) /\oo/\
#
# Share and enjoy! :-)


# setup
#
export RESET_TSTAMP_VERSION="0.4 2022-04-23"
export USAGE="usage: $0 [-h] [-V] [-v level] [-l limit_ioccc.h]

    -h              print help and exit
    -V              print version and exit
    -v level        set debug level (def: 0)
    -l limit_ioccc.h   limit file (def: ./limit_ioccc.h)

Exit codes:
     0   timestamp updated
     1   A verification phase test failed
     2   -h and help string printed or -V and version string printed
     3   Command line usage error
 >= 10   internal error

$0 version: $RESET_TSTAMP_VERSION"
export V_FLAG="0"
export LIMIT_IOCCC_H="./limit_ioccc.h"

# parse args
#
while getopts :hv:Vl: flag; do
    case "$flag" in
    h) echo "$USAGE" 1>&2
       exit 2
       ;;
    v) V_FLAG="$OPTARG";
       ;;
    V) echo "$RESET_TSTAMP_VERSION"
       exit 2
       ;;
    l) LIMIT_IOCCC_H="$OPTARG";
       ;;
    \?) echo "$0: ERROR: invalid option: -$OPTARG" 1>&2
       exit 3
       ;;
    :) echo "$0: ERROR: option -$OPTARG requires an argument" 1>&2
       exit 3
       ;;
   *)
       ;;
    esac
done
if [[ -z $LIMIT_IOCCC_H ]]; then
    echo "$0: ERROR: -b $LIMIT_IOCCC_H name cannot be empty" 1>&2
    exit 9
fi

# check args
#
shift $(( OPTIND - 1 ));
if [[ $# -gt 0 ]]; then
    echo "$0: ERROR: Expected 0 args, found: $#" 1>&2
    exit 9
fi

# firewall
#
if [[ ! -e $LIMIT_IOCCC_H ]]; then
    echo "$0: ERROR: limit_ioccc.h file not found: $LIMIT_IOCCC_H" 1>&2
    exit 7
fi
if [[ ! -f $LIMIT_IOCCC_H ]]; then
    echo "$0: ERROR: limit_ioccc.h not a regular file: $LIMIT_IOCCC_H" 1>&2
    exit 7
fi
if [[ ! -r $LIMIT_IOCCC_H ]]; then
    echo "$0: ERROR: limit_ioccc.h not a readable file: $LIMIT_IOCCC_H" 1>&2
    exit 7
fi
if [[ ! -w $LIMIT_IOCCC_H ]]; then
    echo "$0: ERROR: limit_ioccc.h not a writable file: $LIMIT_IOCCC_H" 1>&2
    exit 7
fi

# debug
#
if [[ $V_FLAG -ge 5 ]]; then
    echo "$0: debug[5]: RESET_TSTAMP_VERSION=$RESET_TSTAMP_VERSION" 1>&2
    echo "$0: debug[5]: LIMIT_IOCCC_H=$LIMIT_IOCCC_H" 1>&2
fi

# verify that we have the rpl command
#
RPL_CMD=$(type -P rpl)
if [[ -z "$RPL_CMD" ]]; then
    echo "$0: ERROR: rpl not found" 1>&2
    echo "$0: ERROR: If do not have the rpl tool, then you may not perform this action." 1>&2
    exit 1
fi
GDATE_CMD=$(type -P gdate)
if [[ -z "$GDATE_CMD" ]]; then
    echo "$0: ERROR: date not found" 1>&2
    echo "$0: ERROR: If do not have the rpl tool, then you may not perform this action." 1>&2
    exit 1
fi

# Phase 0 of verification
#
echo
echo 'Yes, we make it very hard to run this script for good reason.'
echo 'Only IOCCC judges can perform the ALL the steps needed to complete this action.'
echo
echo 'WARNING: This script will invalidate all timestamps prior to now.'
echo 'WARNING: Only run this script when you wish to invalidate all timestamps'
echo 'WARNING: because you made a fundamental change to a critical JSON format,'
echo 'WARNING: made a fundamental change to the compressed tarball file structure,'
echo 'WARNING: or made a critical change to limit_ioccc.h that is MORE restrictive.'
echo
echo 'WARNING: DO NOT run this script simply for a new IOCCC!'
echo
echo 'WARNING: If you wish to do this, please enter the following phrase:'
echo
echo '    Please invalidate all IOCCC timestamps prior to now'
echo
echo -n 'Enter the phrase: '
read -r ANSWER
if [[ "$ANSWER" != 'Please invalidate all IOCCC timestamps prior to now' ]]; then
    echo "$0: notice: Wise choice, MIN_TIMESTAMP was not changed." 1>&2
    exit 2
fi

# Phase 1 of verification
#
echo
echo 'WARNING: Are you sure you want to invalidate all existing compressed tarballs'
echo 'WARNING: and all prior .info.JSON files that have been made,'
echo 'WARNING: and all prior .author.JSON files that have been made,'
echo 'WARNING: forcing everyone to rebuild their compressed tarball entries?'
echo
echo 'WARNING: If you really wish to do this, please enter the following phrase:'
echo
echo '    I understand this'
echo
echo 'WARNING: followed by the phrase:'
echo
echo '    and apologise to those with existing IOCCC compressed tarballs'
echo
echo -n 'Enter that phrase: '
read -r ANSWER
if [[ "$ANSWER" != 'I understand this and apologise to those with existing IOCCC compressed tarballs' ]]; then
    echo "$0: notice: Wise choice, MIN_TIMESTAMP was not changed." 1>&2
    exit 3
fi

# Phase 2 of verification
#
echo
echo 'Enter the phrase that is required (even if you are from that very nice place).'
echo
echo -n 'Enter the required phrase: '
read -r ANSWER
if [[ "$ANSWER" != 'Sorry (tm Canada) :=)' ]]; then
    echo "$0: notice: Wise choice, MIN_TIMESTAMP was not changed." 1>&2
    exit 4
fi

# Phase 3 of verification
#
echo
echo -n 'WARNING: Enter the existing value of MIN_TIMESTAMP: '
read -r OLD_MIN_TIMESTAMP
FORMED_OLD_NOW="$($GDATE_CMD -u --date=@"$OLD_MIN_TIMESTAMP" '+%a %b %d %H:%M:%S %Y UTC')"
if [[ -z $FORMED_OLD_NOW ]]; then
    echo "$0: ERROR: cannot convert OLD_MIN_TIMESTAMP to time string" 1>&2
    exit 5
fi
export FORMED_OLD_NOW
export NOW
NOW="$(date '+%s')"
if grep -q "$OLD_MIN_TIMESTAMP" "$LIMIT_IOCCC_H" >/dev/null 2>&1; then
    echo
    echo 'We guess that you do really really do want to change MIN_TIMESTAMP.'
    echo 'Enter Y if you really want to change the MIN_TIMESTAMP!'
    echo
else
    echo "$0: ERROR: Invalid value of MIN_TIMESTAMP" 1>&2
    exit 5
fi
FORMED_NOW="$($GDATE_CMD -u --date=@"$NOW" '+%a %b %d %H:%M:%S %Y UTC')"
if [[ -z $FORMED_NOW ]]; then
    echo "$0: ERROR: cannot convert MIN_TIMESTAMP to time string" 1>&2
    exit 5
fi
export FORMED_NOW

# Change the timestamp
#
"$RPL_CMD" -w -p \
   "#define MIN_TIMESTAMP ((time_t)$OLD_MIN_TIMESTAMP)" \
   "#define MIN_TIMESTAMP ((time_t)$NOW)" "$LIMIT_IOCCC_H"
status="$?"
if [[ $status -ne 0 ]]; then
    echo "$0: ERROR: rpl failed to modify $LIMIT_IOCCC_H: error code: $status" 1>&2
    exit 6
fi
echo
echo "FYI:"
echo
echo "OLD_MIN_TIMESTAMP=$OLD_MIN_TIMESTAMP"
echo "FORMED_OLD_NOW=$FORMED_OLD_NOW"
echo "NOW=$NOW"
echo "FORMED_NOW=$FORMED_NOW"
echo
echo "This line in $LIMIT_IOCCC_H as it exists now, is:"
echo
grep '^#define MIN_TIMESTAMP' "$LIMIT_IOCCC_H"
echo
echo "$0: notice: You still need to:"
echo
echo '    make clobber all test'
echo
echo 'then:'
echo
echo "	  make ./limit_ioccc.sh"
echo "    ./vermod.sh -v 1 -n -Q $OLD_MIN_TIMESTAMP $NOW"
echo "    $RPL_CMD -s -x'.json' -R -- '$FORMED_OLD_NOW' '$FORMED_NOW' ./test_JSON"
echo
echo 'If all is well, then:'
echo
echo "    ./vermod.sh -v 1 -Q $OLD_MIN_TIMESTAMP $NOW"
echo "    $RPL_CMD -x'.json' -R -- '$FORMED_OLD_NOW' '$FORMED_NOW' ./test_JSON"
echo
echo "$0: notice: And if all is well, commit and push the change to the GitHub repo!"
echo

# All Done!!! -- Jessica Noll, Age 2
#
exit 0
